test/e2e_node/container_lifecycle_test.go

```diff
@@ -886,6 +887,355 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifecycle", func() {
+
+			ginkgo.By("Parsing results")
+			pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+			results := parseOutput(ctx, f, pod)
+
+			ginkgo.By("Analyzing results")
+			framework.ExpectNoError(results.StartsBefore(init1, init2))
+			framework.ExpectNoError(results.ExitsBefore(init1, init2))
+
+			framework.ExpectNoError(results.StartsBefore(init2, init3))
+			framework.ExpectNoError(results.ExitsBefore(init2, init3))
+
+			gomega.Expect(pod.Status.InitContainerStatuses[0].RestartCount).To(gomega.Equal(int32(0)))
+			gomega.Expect(pod.Status.InitContainerStatuses[1].RestartCount).To(gomega.Equal(int32(0)))
+			gomega.Expect(pod.Status.InitContainerStatuses[2].RestartCount).To(gomega.Equal(int32(0)))
+		})
+
+		ginkgo.It("should not restart any completed init container, even after the completed init container statuses have been removed and the kubelet restarted", func(ctx context.Context) {
+			ginkgo.By("stopping the kubelet")
+			startKubelet := stopKubelet()
+			// wait until the kubelet health check will fail
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeFalseBecause("kubelet should be stopped"))
+
+			ginkgo.By("removing the completed init container statuses from the container runtime")
+			rs, _, err := getCRIClient()
+			framework.ExpectNoError(err)
+
+			pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+			for _, c := range pod.Status.InitContainerStatuses {
+				if c.State.Terminated == nil || c.State.Terminated.ExitCode != 0 {
+					continue
+				}
+
+				tokens := strings.Split(c.ContainerID, "://")
+				gomega.Expect(tokens).To(gomega.HaveLen(2))
+
+				containerID := tokens[1]
+
+				err := rs.RemoveContainer(ctx, containerID)
+				framework.ExpectNoError(err)
+			}
+
+			ginkgo.By("restarting the kubelet")
+			startKubelet()
+			// wait until the kubelet health check will succeed
+			gomega.Eventually(ctx, func() bool {
+				return kubeletHealthCheck(kubeletHealthCheckURL)
+			}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be restarted"))
+
+			ginkgo.By("ensuring that no completed init container is restarted")
+			gomega.Consistently(ctx, func() bool {
+				pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err)
+				for _, status := range pod.Status.InitContainerStatuses {
+					if status.State.Terminated == nil || status.State.Terminated.ExitCode != 0 {
+						continue
+					}
+
+					if status.RestartCount > 0 {
+						return false
+					}
+				}
+				return true
+			}, 1*time.Minute, f.Timeouts.Poll).Should(gomega.BeTrueBecause("no completed init container should be restarted"))
+
+			ginkgo.By("Analyzing results")
+			// Cannot analyze the results with the container logs as the
```

Comment:
Maybe we should test the behavior mentioned in the comment. For example, check if there are no startup logs for initContainers removed from the container runtime, while the startup logs for regular containers are still there. Right now, if the container status can't be found, it looks like the logs we get have an error message saying `"unable to retrieve container logs for <containerID>"`. This messes up the log parsing.

I'm not sure if this is the right way to go, but how about changing it so that the error message doesn't get written to the logs when the container status can't be found? You can see a similar handing [here](https://github.com/kubernetes/kubernetes/blob/v1.30.2/test/e2e/storage/podlogs/podlogs.go#L219) as well. So, something like this:

```diff
diff --git a/test/e2e/framework/pod/resource.go b/test/e2e/framework/pod/resource.go
index 3039a9bc821..ca96aa374cf 100644
--- a/test/e2e/framework/pod/resource.go
+++ b/test/e2e/framework/pod/resource.go
@@ -474,6 +474,9 @@ func getPodLogsInternal(ctx context.Context, c clientset.Interface, namespace, p
        if strings.Contains(string(logs), "Internal Error") {
                return "", fmt.Errorf("Fetched log contains \"Internal Error\": %q", string(logs))
        }
+       if strings.Contains(string(logs), "unable to retrieve container logs for ") {
+               return "", fmt.Errorf("Missing container status for %q: %q", containerName, string(logs))
+       }
        return string(logs), err
 }
```

Then, in the test case, we'll verify that only the regular container is starting.

```diff
diff --git a/test/e2e_node/container_lifecycle_test.go b/test/e2e_node/container_lifecycle_test.go
index 8dcb758bbab..895f49e3fa0 100644
--- a/test/e2e_node/container_lifecycle_test.go
+++ b/test/e2e_node/container_lifecycle_test.go
@@ -1051,9 +1051,19 @@ var _ = SIGDescribe(framework.WithSerial(), "Containers Lifec
ycle", func() {
                                return true
                        }, 1*time.Minute, f.Timeouts.Poll).Should(gomega.BeTrueBecau
se("no completed init container should be restarted"))

+                       ginkgo.By("Parsing results")
+                       pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+                       framework.ExpectNoError(err)
+                       results := parseOutput(ctx, f, pod)
+
                        ginkgo.By("Analyzing results")
-                       // Cannot analyze the results with the container logs as the
-                       // container statuses have been removed from container runtime.
+                       ginkgo.By("Verifying that only the regular container starts because the statues of init containers have been removed from container runtime")
+                       framework.ExpectNoError(results.DoesntStart(init1))
+                       framework.ExpectNoError(results.DoesntStart(init2))
+                       framework.ExpectNoError(results.DoesntStart(init3))
+                       framework.ExpectNoError(results.Starts(regular1))
+
+                       ginkgo.By("Verifying that the init containers have not been restarted")
                        gomega.Expect(pod.Status.InitContainerStatuses[0].RestartCount).To(gomega.Equal(int32(0)))
                        gomega.Expect(pod.Status.InitContainerStatuses[1].RestartCount).To(gomega.Equal(int32(0)))
                        gomega.Expect(pod.Status.InitContainerStatuses[2].RestartCount).To(gomega.Equal(int32(0)))
```

Reply:
I am not sure checking the unstructured error message here is a good idea. We will see how other reviewers' comments.

Anyway, fixed it to check the regular1 is running with podStatus, thanks.